File: src/chapter02/GenericList.java
class Item {
    private Object value;
    private Item next;

    public Item(Object value) {
        this.value = value;
    }

    public Object getValue() {
        return value;
    }

    public Item getNext() {
        return next;
    }

    public void setNext(Item next) {
        this.next = next;
    }
}

interface Printer {
    public void print(Object value);
}

class LinePrinter implements Printer {
    public void print(Object value) {
        System.out.println(value);
    }
}

class List {
    private Item head;
    private int n;

    public List(Object value) {
        head = new Item(value);
        n = 1;
    }

    public Item getHead() {
        return head;
    }

    public void append(Object value) {
        Item newItem = new Item(value);
        for (Item tmp = head; tmp != null; tmp = tmp.getNext()) {
            if (tmp.getNext() == null) {
                tmp.setNext(newItem);
                n++;
                return;
            }
        }
    }

    public void each(Printer printer) {
        for (Item tmp = head; tmp != null; tmp = tmp.getNext()) {
            printer.print(tmp.getValue());
        }
    }

    public int size() {
        return n;
    }

    public Item get(int i) {
        Item item = null;

        if (i < 0 || i >= n) {
            return null;
        }

        for (item = head; item != null; item = item.getNext()) {
            if (i == 0) {
                return item;
            }
            i--;
        }

        return null;
    }
}

class GenericList {
    public static void main(String args[]) {
        List list = new List("everything");
        list.append("that");
        list.append("begins");
        list.append("also");
        list.append("ends");
        list.each(new LinePrinter());
        System.out.println(list.size() + " words");
        System.out.println("first word: " + list.get(0).getValue());
        System.out.println("second word: " + list.get(1).getValue());
        System.out.println("third word: " + list.get(2).getValue());
        System.out.println("fourth word: " + list.get(3).getValue());
        System.out.println("fifth word: " + list.get(4).getValue());
    }
}

File: src/chapter02/genericlist.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "genericlist.h"

#define STRLEN 20

void printFloat(Item *item)
{
    printf("%.2f\n", *(float*)item->value);
}

void printString(Item *item)
{
    printf("%s\n", (char*)item->value);
}

int main()
{
    List *scores, *names;
    float first, second, third;
    char monkey[STRLEN], donkey[STRLEN], honkey[STRLEN];
    int n_scores, n_names;

    first = 3.14;
    second = 2.88;
    third = 7.42;

    strncpy(monkey, "San Pellegrino", STRLEN);
    strncpy(donkey, "Kristoffer", STRLEN);
    strncpy(honkey, "Hasler", STRLEN);

    scores = new_list((void*)&first);
    append(scores, (void*)&second);
    append(scores, (void*)&third);

    names = new_list((void*)&monkey);
    append(names, (void*)donkey);
    append(names, (void*)honkey);

    each(scores, printFloat);
    each(names, printString);

    n_scores = size(scores);
    n_names = size(scores);

    printf("%d scores\n", n_scores);
    printf("%d names\n", n_names);

    printf("second score: %f\n", *(float*)get(scores, 1));
    printf("third name: %s\n", (char*)get(names, 2));

    del_list(scores);
    del_list(names);

    return 0;
}

List *new_list(void *value)
{
    List *list;
    Item *item;

    item = (Item*)malloc(sizeof(Item));
    item->value = value;
    item->next = NULL;

    list = (List*)malloc(sizeof(List));
    list->head = item;

    return list;
}

void del_list(List *list)
{
    Item *tmp, *last;

    last = NULL;
    for (tmp = list->head; tmp != NULL; tmp = tmp->next) {
        if (last != NULL) {
            free(last);
        }
        last = tmp;
    }
    if (last != NULL) {
        free(last);
    }

    free(list);
}

void append(List *list, void *value)
{
    Item *new_item, *tmp;
    
    new_item = (Item*)malloc(sizeof(Item));
    new_item->value = value;
    new_item->next = NULL;

    for (tmp = list->head; tmp != NULL; tmp = tmp->next) {
        if (tmp->next == NULL) {
            tmp->next = new_item;
            return;
        }
    }
}

void each(List *list, void (*func)(Item*))
{
    Item *tmp;

    for (tmp = list->head; tmp != NULL; tmp = tmp->next) {
        func(tmp);
    }
}

int size(List *list)
{
    Item *item = NULL;
    int n = 0;

    for (item = list->head; item != NULL; item = item->next) {
        n++;
    }

    return n;
}

void *get(List *list, int i)
{
    Item *item = NULL;
    int n;

    n = size(list);
    if (i < 0 || i >= n) {
        return NULL;
    }

    for (item = list->head; item != NULL; item = item->next) {
        if (i == 0) {
            return item->value;
        }
        i--;
    }

    return NULL;
}
